### PR TITLE
criu-ns: Fix --restore-detached option

### LINKED ISSUE #807: Impossible to restore any process due to PID increment rates.
How to reproduce:
1. Dump a process with lots of RAM usage and many threads (like ZPAQ).
2. Reboot.
3. When trying to restore, it fails with "Pid $PID do not match expected $PID".

The `unshare` method does not apply since it does not cover shell jobs.
The problem is that many thing in the OS creates new processes frequently and take up the PID desired.
They create new processes every few seconds, and they are too important (or simply impossible) to kill.
For example, "systemd-udevd" now frequently ends and spawns child processes after an update.
BTRFS also frequently creates kernel threads.
Even worse, CRIU now set the starting point of new PIDs to the restored process' PID, so no matter how many times you try, it will always fail because the next few PIDs are quickly taken.
I think that a workaround is to create empty threads in advance so the PIDs won't be taken during restoration, but I am not sure.

This is a fundamental problem. You tried to restore processes to their original PIDs, which is impossible in reality.

### LINKED ISSUE #1278: target process failed to start after  criu-ns restore  with “-d”
I used criu-ns to restore with "-d", while target process didnot restart  after criu detach  as  follwing log.

#./criu-ns  restore  -j -D ./graph500-ckpt/ --tcp-established -d -v4

Whether criu-ns  cannot work with "-d"?

(01.380027) 73655 (native) is going to execute the syscall 215, required is 215
(01.380061) 73655 was stopped
(01.380087) 73647 was trapped
(01.380092) 73647 (native) is going to execute the syscall 215, required is 215
(01.380127) 73647 was stopped
(01.380153) 73636 was trapped
(01.380159) 73636 (native) is going to execute the syscall 215, required is 215
(01.380192) 73636 was stopped
(01.380218) 73625 was trapped
(01.380223) 73625 (native) is going to execute the syscall 215, required is 215
(01.380256) 73625 was stopped
(01.380282) 73615 was trapped
(01.380288) 73615 (native) is going to execute the syscall 215, required is 215
(01.380319) 73615 was stopped
(01.380345) 73607 was trapped
(01.380350) 73607 (native) is going to execute the syscall 215, required is 215
(01.380384) 73607 was stopped
(01.380408) 73605 was trapped
(01.380414) 73605 (native) is going to execute the syscall 215, required is 215
(01.380456) 73605 was stopped
(01.380481) 73601 was trapped
(01.380487) 73601 (native) is going to execute the syscall 215, required is 215
(01.380520) 73601 was stopped
(01.380528) Running pre-resume scripts
(01.381344) Restore finished successfully. Tasks resumed.
(01.381354) Writing stats
(01.381672) Running post-resume scripts

### PR DESCRIPTION
**Pull request by rst0git**
The `criu-ns` script creates a new PID namespace where CRIU runs as the "init" process.

When using the `--restore-detached` option with `criu-ns`, users expect `criu-ns` to exit without killing the restored process tree. Thus, `criu-ns` should exit instead of waiting for CRIU's status and it should not pass the `--restore-detached` to CRIU to prevent terminating the init process.

Resolves #1278
Resolves #807

### CODE REVIEW

File: scripts/criu-ns [hunk 1]
```diff
@@ -61,39 +61,87 @@ def run_criu():
...
+	Create an init process to keep the new PID namespace
+	after CRIU exits.
+	"""
+	(r_pipe, w_pipe) = os.pipe()
+	# Spawn the init
+	if os.fork() == 0:
+		os.close(r_pipe)
+		_mount_new_proc()
+		# Spawn CRIU binary
+		criu_pid = os.fork()
```
Comment (avagin) [inline scripts/criu-ns]: Hmmm. I think you are going to reintroduce the bug that you fixed a month ago:
https://github.com/checkpoint-restore/criu/commit/265569a40f04326dc2b310eb5ddbac9c1cb989e3

Comment (rst0git) [reply scripts/criu-ns]: As far as understand when using `--restore-detached`, CRIU exits after restore, therefore we need another process to keep running within the namespace as PID=1 after restore. As illustrated in https://criu.org/Tree_after_restore#Detached

When CRIU is used without `--restore-detached`, it will run as PID=1.

File: scripts/criu-ns [hunk 2]
```diff
@@ -18,190 +18,238 @@ MS_SLAVE = 1 << 19
...
-	os.execlp('criu', *['criu'] + sys.argv[1:])
+    print(sys.argv)
+    os.execlp('criu', *['criu'] + sys.argv[1:])
+
+
+def _mount_new_proc():
+    """
+    Mount new /proc filesystem.
+    """
+    if _mount(None, b"/", None, MS_SLAVE|MS_REC, None):
```
Comment (Snorch) [inline scripts/criu-ns:68]: Why not MS_PRIVATE? why do you need MS_SLAVE, for which exact propagation?

File: scripts/criu-ns (hunk 1 above)
Comment (Snorch) [reply scripts/criu-ns]: Probably https://github.com/checkpoint-restore/criu/commit/265569a40f04326dc2b310eb5ddbac9c1cb989e3 could've been fixed with just forking criu with some unused pid which will not intersect with restored pids, and this way we would've hit two birds with one stone.

File: scripts/criu-ns (hunk 2 above)
Comment (Snorch) [reply scripts/criu-ns:68]: I'm not strictly against it as it's code copying, but this propagation is useless FMPOV.

Comment (avagin) [reply scripts/criu-ns]: If we create a mount namespace just for mounting /proc, MS_SLAVE is the best choice. All changes in the host mount namespace will be propagated into this namespace.

Comment (Snorch) [reply scripts/criu-ns:68]: > If we create a mount namespace just for mounting /proc, MS_SLAVE is the best choice. All changes in the host mount namespace will be propagated into this namespace.

*offtop*

@avagin Note for the case of "umount a file system and detach its device from the system" probably a more robust way would be not to rely on propagation but to create some new syscall umount3 (or smth. like this) to just umount all mounts of the superblock.

It would probably also help with autofs problems: now when autofs service is restarted it tries to umount autofs and mount a new one in it's place, if you clone mountns and make it rprivate, bindmount of old autofs mount in it will remain but be in some kind of broken state. This new syscall can give an ability to umount all kinds of leftover mounts.

File: scripts/criu-ns (hunk 1 above)
Comment (criupatchwork) [reply scripts/criu-ns]: @rst0git could you explain how you resolved this comment?

Comment (rst0git) [reply scripts/criu-ns]: As reported in https://github.com/checkpoint-restore/criu/issues/1278, `-d` when used with `criu-ns` is currently broken. It does not work because CRIU exists after restore to detach itself from the process tree, and if  CRIU runs as PID=1 all processes in the PID namespace will be terminated.
This issue was introduced with https://github.com/checkpoint-restore/criu/commit/265569a40f04326dc2b310eb5ddbac9c1cb989e3. 

> could you explain how you resolved this comment?

To fix the broken `-d` functionality, I'm reusing the previous code/behaviour before the commit mention above. Although it has the limitation mentioned in this comment, it works.

In other words, I haven't (yet) resolved the problem mentioned in this comment. It remains a limitation of `criu-ns`.

File: scripts/criu-ns [hunk 3]
```diff
@@ -18,190 +18,195 @@ MS_SLAVE = 1 << 19
...
+        raise OSError(_errno, errno.errorcode[_errno])
 
-	return (None, None)
+    restore_detached = False
+    if '-d' in restore_args:
+        restore_detached = True
+        restore_args.remove('-d')
+    if '--restore-detached' in restore_args:
+        restore_detached = True
+        restore_args.remove('--restore-detached')
```
Comment (Snorch) [inline scripts/criu-ns:101]: A bit funny that to restore "detached" we remove "--restore-detached". =)

**Review by Snorch — APPROVED** [review]
This looks good for me!

File: scripts/criu-ns (hunk 3 above)
Comment (rst0git) [reply scripts/criu-ns:101]: Yes, it is a bit funny and paradoxical, but I hope the commit message is sufficient to explain why.

**Comment by rst0git** [comment]
@Snorch Thank you for the code review!

**Comment by rst0git** [comment]
Thanks!
